File: apps/worker/src/aurum_worker/models/positions.py
"""Read-side Position state with no broker mutation capability."""

from __future__ import annotations

from enum import StrEnum
from typing import Literal
from uuid import UUID

from pydantic import AwareDatetime, model_validator

from aurum_worker.models.base import (
    FiniteFloat,
    Identifier,
    PositiveFloat,
    PositiveInt,
    WireModel,
)
from aurum_worker.models.trading import PermittedVolume, TradeDirection


class PositionStatus(StrEnum):
    OPEN = "open"
    CLOSE_REQUESTED = "close_requested"
    CLOSING = "closing"
    CLOSED = "closed"
    MISMATCH = "mismatch"


class Position(WireModel):
    id: UUID
    position_version: PositiveInt
    user_id: UUID
    trading_account_id: UUID
    account_type: Literal["demo"]
    canonical_symbol: Literal["XAUUSD"]
    direction: TradeDirection
    volume: PermittedVolume
    entry: PositiveFloat
    current: PositiveFloat
    stop_loss: PositiveFloat
    take_profit: PositiveFloat
    unrealized_pnl: FiniteFloat
    r_multiple: FiniteFloat
    status: PositionStatus
    opened_at: AwareDatetime
    updated_at: AwareDatetime
    closed_at: AwareDatetime | None

    @model_validator(mode="after")
    def validate_position_state(self) -> Position:
        if self.direction is TradeDirection.BUY:
            prices_are_ordered = self.stop_loss < self.entry < self.take_profit
        else:
            prices_are_ordered = self.take_profit < self.entry < self.stop_loss
        if not prices_are_ordered:
            raise ValueError(
                "Entry, Stop Loss, and Take Profit are invalid for direction."
            )

        if self.status is PositionStatus.CLOSED and self.closed_at is None:
            raise ValueError("A closed Position requires a close timestamp.")
        if self.status is not PositionStatus.CLOSED and self.closed_at is not None:
            raise ValueError("Only a closed Position may carry a close timestamp.")
        if self.closed_at is not None and self.closed_at < self.opened_at:
            raise ValueError("A Position cannot close before it opened.")
        return self


class PersistedPosition(WireModel):
    """SQL-aligned row after snake_case has been mapped to wire aliases."""

    id: UUID
    owner_id: UUID
    trading_account_id: UUID
    trade_proposal_id: UUID
    broker_order_id: UUID
    broker_position_reference: Identifier
    position_version: PositiveInt
    direction: TradeDirection
    volume: PermittedVolume
    entry_price: PositiveFloat
    current_price: PositiveFloat
    stop_loss_price: PositiveFloat
    take_profit_price: PositiveFloat
    unrealized_pnl: FiniteFloat
    r_multiple: FiniteFloat
    status: PositionStatus
    opened_at: AwareDatetime
    closed_at: AwareDatetime | None
    created_at: AwareDatetime
    updated_at: AwareDatetime

    def to_position(self) -> Position:
        return Position.model_validate(
            {
                "id": self.id,
                "positionVersion": self.position_version,
                "userId": self.owner_id,
                "tradingAccountId": self.trading_account_id,
                "accountType": "demo",
                "canonicalSymbol": "XAUUSD",
                "direction": self.direction,
                "volume": self.volume,
                "entry": self.entry_price,
                "current": self.current_price,
                "stopLoss": self.stop_loss_price,
                "takeProfit": self.take_profit_price,
                "unrealizedPnl": self.unrealized_pnl,
                "rMultiple": self.r_multiple,
                "status": self.status,
                "openedAt": self.opened_at,
                "updatedAt": self.updated_at,
                "closedAt": self.closed_at,
            }
        )

File: packages/contracts/src/positions.ts
import { z } from "zod";

import {
  FiniteNumberSchema,
  IsoDateTimeSchema,
  PositiveNumberSchema,
  UuidSchema,
} from "./primitives";

export const POSITION_STATUSES = [
  "open",
  "close_requested",
  "closing",
  "closed",
  "mismatch",
] as const;

export const PositionStatusSchema = z.enum(POSITION_STATUSES);
export type PositionStatus = z.infer<typeof PositionStatusSchema>;

/** Read-side Position state. This contract grants no broker mutation capability. */
export const PositionSchema = z
  .object({
    id: UuidSchema,
    positionVersion: z.number().int().positive(),
    userId: UuidSchema,
    tradingAccountId: UuidSchema,
    accountType: z.literal("demo"),
    canonicalSymbol: z.literal("XAUUSD"),
    direction: z.enum(["BUY", "SELL"]),
    volume: PositiveNumberSchema.max(0.01),
    entry: PositiveNumberSchema,
    current: PositiveNumberSchema,
    stopLoss: PositiveNumberSchema,
    takeProfit: PositiveNumberSchema,
    unrealizedPnl: FiniteNumberSchema,
    rMultiple: FiniteNumberSchema,
    status: PositionStatusSchema,
    openedAt: IsoDateTimeSchema,
    updatedAt: IsoDateTimeSchema,
    closedAt: IsoDateTimeSchema.nullable(),
  })
  .strict()
  .superRefine((position, context) => {
    const pricesAreOrdered =
      position.direction === "BUY"
        ? position.stopLoss < position.entry &&
          position.entry < position.takeProfit
        : position.takeProfit < position.entry &&
          position.entry < position.stopLoss;
    if (!pricesAreOrdered) {
      context.addIssue({
        code: "custom",
        path: ["stopLoss"],
        message: "Entry, Stop Loss, and Take Profit are invalid for direction.",
      });
    }

    if (position.status === "closed" && position.closedAt === null) {
      context.addIssue({
        code: "custom",
        path: ["closedAt"],
        message: "A closed Position requires a close timestamp.",
      });
    }
    if (position.status !== "closed" && position.closedAt !== null) {
      context.addIssue({
        code: "custom",
        path: ["closedAt"],
        message: "Only a closed Position may carry a close timestamp.",
      });
    }
    if (
      position.closedAt !== null &&
      Date.parse(position.closedAt) < Date.parse(position.openedAt)
    ) {
      context.addIssue({
        code: "custom",
        path: ["closedAt"],
        message: "A Position cannot close before it opened.",
      });
    }
  });

export type Position = z.infer<typeof PositionSchema>;

/** SQL-aligned Position row after snake_case has been mapped to camelCase. */
export const PersistedPositionSchema = z
  .object({
    id: UuidSchema,
    ownerId: UuidSchema,
    tradingAccountId: UuidSchema,
    tradeProposalId: UuidSchema,
    brokerOrderId: UuidSchema,
    brokerPositionReference: z.string().trim().min(1).max(160),
    positionVersion: z.number().int().positive(),
    direction: z.enum(["BUY", "SELL"]),
    volume: PositiveNumberSchema.max(0.01),
    entryPrice: PositiveNumberSchema,
    currentPrice: PositiveNumberSchema,
    stopLossPrice: PositiveNumberSchema,
    takeProfitPrice: PositiveNumberSchema,
    unrealizedPnl: FiniteNumberSchema,
    rMultiple: FiniteNumberSchema,
    status: PositionStatusSchema,
    openedAt: IsoDateTimeSchema,
    closedAt: IsoDateTimeSchema.nullable(),
    createdAt: IsoDateTimeSchema,
    updatedAt: IsoDateTimeSchema,
  })
  .strict();

export type PersistedPosition = z.infer<typeof PersistedPositionSchema>;

export function positionFromPersisted(record: PersistedPosition): Position {
  return PositionSchema.parse({
    id: record.id,
    positionVersion: record.positionVersion,
    userId: record.ownerId,
    tradingAccountId: record.tradingAccountId,
    accountType: "demo",
    canonicalSymbol: "XAUUSD",
    direction: record.direction,
    volume: record.volume,
    entry: record.entryPrice,
    current: record.currentPrice,
    stopLoss: record.stopLossPrice,
    takeProfit: record.takeProfitPrice,
    unrealizedPnl: record.unrealizedPnl,
    rMultiple: record.rMultiple,
    status: record.status,
    openedAt: record.openedAt,
    updatedAt: record.updatedAt,
    closedAt: record.closedAt,
  });
}
